### PR TITLE
dispatch Series[datetime64] ops to DatetimeIndex

### LINKED ISSUE #17837: diff of datetime column with NaT does not result into TimeDelta
#### Code Sample, a copy-pastable example if possible

```python
In[40]: df = pd.DataFrame([[0]], dtype="datetime64[ns]")

In[46]: df[0] - pd.to_datetime("nat")
Out[47]: 
0   NaT
Name: 0, dtype: datetime64[ns]

In[47]: df[0] - pd.to_datetime("2017")
Out[48]: 
0   -17167 days
Name: 0, dtype: timedelta64[ns]

```
#### Problem description

Operation - on datetimes should always return a timedelta.

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.2.final.0
python-bits: 64
OS: Linux
OS-release: 4.10.0-35-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8
pandas: 0.20.3
pytest: None
pip: 9.0.1
setuptools: 36.2.7
Cython: None
numpy: 1.13.3
scipy: 0.19.0
xarray: None
IPython: 6.0.0
sphinx: None
patsy: None
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.0.1
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 0.999999999
sqlalchemy: 1.1.14
pymysql: None
psycopg2: 2.7.3 (dt dec pq3 ext lo64)
jinja2: 2.9.6
s3fs: None
pandas_gbq: None
pandas_datareader: None

</details>

### LINKED ISSUE #18808: Series.__sub__(NaT) vs DatetimeIndex.__sub__(NaT)
One treats NaT result as datetime, other treats it as timedelta.

```
>>> dti  = pd.DatetimeIndex([pd.NaT, pd.Timestamp('19900315')])
>>> ser = pd.Series(dti)

>>> dti - pd.NaT
TimedeltaIndex([NaT, NaT], dtype='timedelta64[ns]', freq=None)

>>> ser - pd.NaT
0   NaT
1   NaT
dtype: datetime64[ns]
```

### LINKED ISSUE #18850: Series[datetime64] - PeriodIndex vs DatetimeIndex - PeriodIndex
`PeriodIndex` w/ `Series` is allowed, `PeriodIndex` w/ `DatetimeIndex` is not.  Is this intentional?  If not, which should be made to match the other?

```
>>> dti = pd.date_range('2016-01-01', periods=3)
>>> ser = pd.Series(dti)
>>> pi = dti.to_period()

>>> ser - pi
0   0 days
1   0 days
2   0 days
dtype: timedelta64[ns]

>>> dti - pi
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pandas/core/indexes/datetimelike.py", line 724, in __sub__
    typ2=type(other).__name__))
TypeError: cannot subtract DatetimeIndex and PeriodIndex

>>> pi - ser
0   0 days
1   0 days
2   0 days
dtype: timedelta64[ns]

>>> pi - dti
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pandas/core/indexes/datetimelike.py", line 731, in __rsub__
    return -(self - other)
  File "pandas/core/indexes/datetimelike.py", line 724, in __sub__
    typ2=type(other).__name__))
TypeError: cannot subtract DatetimeIndex and PeriodIndex
```

### PR DESCRIPTION
**Pull request by jbrockmendel**
This is the culmination of a bunch of recent work.

- [x] closes #18850, closes #18808, closes #17837 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

If merged, this will subsume #18960 and will obviate parts of #18964.  This will also fix (but not test, so we'll leave the issue open for now) the lack of overflow checks #12534.  Also in a follow-up we'll be able to remove a bunch of _TimeOp.

### CODE REVIEW

**Comment by jreback** [comment]
rebase and push again, fixed some hanging by Travis CI

File: pandas/core/ops.py [hunk 1]
```diff
@@ -744,6 +744,13 @@ def wrapper(left, right, name=name, na_op=na_op):
             return NotImplemented
 
         left, right = _align_method_SERIES(left, right)
+        if is_datetime64_dtype(left) or is_datetime64tz_dtype(left):
```
Comment (jreback) [inline pandas/core/ops.py]: and until / unless you want to limit / remove _TimeOp (which is actually ok with me). then this doesn't belong here as I have commented before.

You are welcome to put it *in* _TimeOp for now or rip out TimeOp.

**Comment by jreback** [comment]
I think it will be much cleaner / better to have 1 PR which addresses any/all of TimeOp issues. You seem to want to change it, which is totally fine. But rather than just hack around, let's blow it away.

**Comment by jbrockmendel** [comment]
Yes, I very much want to get rid of `_TimeOp`.  The strategy so far has been to first address the various ways in which `_TimeOp` behaves differently from `DatetimeIndex`/`TimedeltaIndex`, that way we make sure a) we have appropriate tests for each bug that gets fixed, b) appropriate WhatsNew notes, and c) check that these differences are in fact unintentional.

> You are welcome to put it in _TimeOp for now or rip out TimeOp.

For reasons discussed elsewhere, putting it _in_ `_TimeOp` is not a viable option long-term.  What we _can_ do is I can amend this PR to rip out the `is_datetime_lhs` cases, leaving only the `is_timedelta_lhs` cases.  We're a few steps away from being able to rip those out, but we're getting there.

**Comment by jreback** [comment]
that would ls be fine

File: doc/source/whatsnew/v0.23.0.txt [hunk 2]
```diff
@@ -208,6 +208,9 @@ Other API Changes
 - In :func:`read_excel`, the ``comment`` argument is now exposed as a named parameter (:issue:`18735`)
 - Rearranged the order of keyword arguments in :func:`read_excel()` to align with :func:`read_csv()` (:issue:`16672`)
 - The options ``html.border`` and ``mode.use_inf_as_null`` were deprecated in prior versions, these will now show ``FutureWarning`` rather than a ``DeprecationWarning`` (:issue:`19003`)
+- Subtracting ``NaT`` from a :class:`Series` with ``dtype='datetime64[ns]'`` returns a ``Series`` with ``dtype='timedelta64[ns]'`` instead of ``dtype='datetime64[ns]'``(:issue:`18808`)
+- Operations between a :class:`Series` with dtype ``dtype='datetime64[ns]'`` and a :class:`PeriodIndex` will correctly raises ``TypeError`` (:issue:`18850`)
+- Subtraction of :class:`Series` with timezone-aware ``dtype='datetime64[ns]'`` will mis-matched timezones will raise ``TypeError`` instead of ``ValueError`` (issue:`18817`) dtypewith mis-matched timezones will now raise a ``TypeError`` instead of a ``ValueError`` (:issue:`18817`)
```
Comment (jschendel) [inline doc/source/whatsnew/v0.23.0.txt]: looks like some text got duped in the last entry

File: pandas/tests/series/test_timeseries.py [hunk 3]
```diff
@@ -107,7 +107,7 @@ def test_shift(self):
         # incompat tz
         s2 = Series(date_range('2000-01-01 09:00:00', periods=5,
                                tz='CET'), name='foo')
-        pytest.raises(ValueError, lambda: s - s2)
+        pytest.raises(TypeError, lambda: s - s2)
```
Comment (jreback) [inline pandas/tests/series/test_timeseries.py]: is this in the whatsnew notes API changed section?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/series/test_timeseries.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/tests/series/test_timeseries.py]: Yes

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
see how easy that was! removing code is fun!

anyhow obviously want to come back and clean this up even more, and definitly add some comments on what is non-obvious in the code (e.g. where datetimes are dispatched to DTI)

**Pull request merged**